tracer/src/Datadog.Trace/Debugger/Helpers/AsyncHelper.cs

```diff
@@ -0,0 +1,231 @@
+#else
+            return fieldName.Substring(1, indexOfStateMachineSuffix - 1);
+#endif
+        }
+
+        private static bool IsStateMachineField(FieldInfo field)
+        {
+            if (field.Name.StartsWith(StateMachineFieldsPrefix))
+            {
+                AsyncMethodDebuggerInvoker.Log.Information($"Ignoring local (state machine field):  {field.FieldType.Name} {field.DeclaringType?.Name}.{field.Name}");
+                return true;
+            }
+
+            return false;
+        }
+
+        private static bool IsKnownStateMachineField(FieldInfo field)
+        {
+            // builder (various object types), state (int) and "this"
+            if (field.Name is BuilderFieldName or StateFieldName or StateMachineThisFieldName or StateMachineLiveDebuggerAddedField)
+            {
+                return true;
+            }
+
+            // awaiter etc. (e.g. <>u__1)
+            if (field.Name.StartsWith(StateMachineFieldsPrefix))
+            {
+                if (field.FieldType.Name is "TaskAwaiter" or "TaskAwaiter`1" or "ValueTaskAwaiter" or "ValueTaskAwaiter`1")
+                {
+                    return true;
+                }
+                else
+                {
+                    AsyncMethodDebuggerInvoker.Log.Information($"Ignoring local {field.FieldType.Name} {field.DeclaringType?.Name}.{field.Name}");
+                    return true;
+                }
+            }
+
+            return false;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static MethodBase GetAsyncKickoffMethod(Type stateMachineType)
+        {
+            var originalType = stateMachineType.DeclaringType;
+            var allMethods = originalType?.GetMethods(BindingFlags.Instance | BindingFlags.Static | BindingFlags.Public | BindingFlags.NonPublic | BindingFlags.FlattenHierarchy);
+            MethodBase foundMethod = null;
+            for (int i = 0; i < allMethods?.Length; i++)
+            {
+                var currentMethod = allMethods[i];
+                if (currentMethod.GetCustomAttribute<AsyncStateMachineAttribute>()?.StateMachineType == stateMachineType)
+                {
+                    foundMethod = currentMethod;
+                    break;
+                }
+            }
+
+            return foundMethod;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal static object GetAsyncKickoffThisObject<TTarget>(TTarget stateMachine)
+        {
+            var thisField = stateMachine.GetType().GetField(StateMachineThisFieldName, BindingFlags.Instance | BindingFlags.Public);
+            return thisField?.GetValue(stateMachine);
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal static AsyncKickoffMethodInfo GetAsyncKickoffMethodInfo<TTarget>(TTarget stateMachine)
+        {
+            var parent = GetAsyncKickoffThisObject(stateMachine);
```

Comment:
Nitpick: I'd prefer to name `parent` as `kickoffObject` here and to also name `AsyncKickoffMethodInfo.KickoffParentObject` as `KickoffObject`. I feel it'll be clearer as the 'parent' thingy is implementation detail but the meaning is the kickoff-object. It's totally a matter of taste. WDYT?

Reply:
Fixed in 871a652d9b150d63443d87a542e34d4643145adf